docs/userguide/checkpoints.rst

```diff
@@ -1,17 +1,148 @@
+.. _label-memos
+
+Memoization and checkpointing
+-----------------------------
+
+When an app is invoked several times with the same parameters, Parsl can
+reuse the result from the first invocation without executing the app again.
+
+This can save time and computational resources.
+
+This is done in two ways:
+
+* Firstly, *app caching* will allow reuse of results within the same run.
+
+* Building on top of that, *checkpointing* will store results on the filesystem
+  and reuse those results in later runs.
+
+.. _label-appcaching:
+
+App caching
+===========
+
+
+There are many situations in which a program may be re-executed
+over time. Often, large fragments of the program will not have changed 
+and therefore, re-execution of apps will waste valuable time and 
+computation resources. Parsl's app caching solves this problem by 
+storing results from apps that have successfully completed
+so that they can be re-used. 
+
+App caching is enabled by setting the ``cache``
+argument in the :func:`~parsl.app.app.python_app` or :func:`~parsl.app.app.bash_app` 
+decorator to ``True`` (by default it is ``False``). 
+
+.. code-block:: python
+
+   @bash_app(cache=True)
+   def hello (msg, stdout=None):
+       return 'echo {}'.format(msg)
+			
+App caching can be globally disabled by setting ``app_cache=False``
+in the :class:`~parsl.config.Config`.
+
+App caching can be particularly useful when developing interactive programs such as when
+using a Jupyter notebook. In this case, cells containing apps are often re-executed
+during development. Using app caching will ensure that only modified apps are re-executed.
+
+
+App equivalence 
+^^^^^^^^^^^^^^^
+
+Parsl determines app equivalence by storing the a hash
+of the app function. Thus, any changes to the app code (e.g., 
+its signature, its body, or even the docstring within the body)
+will invalidate cached values. 
+Further, Parsl does not traverse imported modules, and thus
```

Comment:
Maybe change "Further" to "However"?

Reply:
i rephrased this whole paragraph and made a new PR #1955